src/llm_models/perplexity.rs

```diff
@@ -0,0 +1,161 @@
+            ],
+        })
+    }
+    ///
+    /// This function leverages Perplexity API to perform any query as per the provided body.
+    ///
+    /// It returns a String the Response object that needs to be parsed based on the self.model.
+    ///
+    async fn call_api(
+        &self,
+        api_key: &str,
+        body: &serde_json::Value,
+        debug: bool,
+    ) -> Result<String> {
+        //Get the API url
+        let model_url = self.get_endpoint();
+
+        //Make the API call
+        let client = Client::new();
+
+        //Send request
+        let response = client
+            .post(model_url)
+            .header(header::CONTENT_TYPE, "application/json")
+            .bearer_auth(api_key)
+            .json(&body)
+            .send()
+            .await?;
+
+        let response_status = response.status();
+        let response_text = response.text().await?;
+
+        if debug {
+            info!(
+                "[debug] Perplexity API response: [{}] {:#?}",
+                &response_status, &response_text
+            );
+        }
+
+        Ok(response_text)
+    }
+
+    //This method attempts to convert the provided API response text into the expected struct and extracts the data from the response
+    fn get_data(&self, response_text: &str, _function_call: bool) -> Result<String> {
+        //Convert API response to struct representing expected response format
+        let completions_response: PerplexityAPICompletionsResponse =
+            serde_json::from_str(response_text)?;
+
+        //Parse the response and return the assistant content
+        completions_response
+            .choices
+            .iter()
+            .filter_map(|choice| choice.message.as_ref())
+            .find(|&message| message.role == Some("assistant".to_string()))
+            .and_then(|message| {
+                message
+                    .content
+                    .as_ref()
+                    .map(|content| sanitize_json_response(content))
+            })
+            .ok_or_else(|| anyhow!("Assistant role content not found"))
+    }
+
+    //This function allows to check the rate limits for different models
+    fn get_rate_limit(&self) -> RateLimit {
+        //Perplexity documentation: https://docs.perplexity.ai/guides/rate-limits
+        RateLimit {
+            tpm: 50 * 127_072, // 50 requests per minute wit max 127,072 context length
+            rpm: 50,           // 50 request per minute
+        }
+    }
+
+    // Accepts a [0-100] percentage range and returns the target temperature based on model ranges
+    fn get_normalized_temperature(&self, relative_temp: u32) -> f32 {
+        // Temperature range documentation: https://docs.perplexity.ai/api-reference/chat-completions
+        // "The amount of randomness in the response, valued between 0 *inclusive* and 2 *exclusive*."
+        let min = 0.0f32;
+        let max = 1.99999f32;
+        map_to_range_f32(min, max, relative_temp)
```

Comment:
Perplexity allows temp to be up to 2 (exclusive), so I had to add a function that accepts an f32 range.